R/CallbackTuning.R

```diff
@@ -0,0 +1,135 @@
+#' @param on_optimization_begin (`function()`)\cr
+#'   Stage called at the beginning of the optimization.
+#'   Called in `Optimizer$optimize()`.
+#'   The function must have two arguments named `callback` and `context`.
+#'   The context available is [bbotk::ContextOptimization].
+#' @param on_optimizer_before_eval (`function()`)\cr
+#'   Stage called after the optimizer proposes points.
+#'   Called in `OptimInstance$eval_batch()`.
+#'   The functions must have two arguments named `callback` and `context`.
+#'   The context available is [bbotk::ContextOptimization].
+#' @param on_eval_after_design (`function()`)\cr
+#'   Stage called after design is created.
+#'   Called in `ObjectiveTuning$eval_many()`.
+#'   The context available is [ContextEval].
+#' @param on_eval_after_benchmark (`function()`)\cr
+#'   Stage called after hyperparameter configurations are evaluated.
+#'   Called in `ObjectiveTuning$eval_many()`.
+#'   The context available is [ContextEval].
+#' @param on_eval_before_archive (`function()`)\cr
+#'   Stage called before performance values are written to the archive.
+#'   Called in `ObjectiveTuning$eval_many()`.
+#'   The context available is [ContextEval].
+#' @param on_optimizer_after_eval (`function()`)\cr
+#'   Stage called after points are evaluated.
+#'   Called in `OptimInstance$eval_batch()`.
+#'   The functions must have two arguments named `callback` and `context`.
+#'   The context available is [bbotk::ContextOptimization].
+#' @param on_result (`function()`)\cr
+#'   Stage called after result are written.
+#'   Called in `OptimInstance$assign_result()`.
+#'   The functions must have two arguments named `callback` and `context`.
+#'   The context available is [bbotk::ContextOptimization].
+#' @param on_optimization_end (`function()`)\cr
+#'   Stage called at the end of the optimization.
+#'   Called in `Optimizer$optimize()`.
+#'   The functions must have two arguments named `callback` and `context`.
+#'   The context available is [bbotk::ContextOptimization].
+#' @param fields (list of `any`)\cr
+#'   List of additional fields.
+#'
+#' @export
+#' @inherit CallbackTuning examples
+callback_tuning = function(id, label = NA_character_, man = NA_character_, on_optimization_begin = NULL, on_optimizer_before_eval = NULL, on_eval_after_design = NULL, on_eval_after_benchmark = NULL, on_eval_before_archive = NULL, on_optimizer_after_eval = NULL, on_result = NULL,  on_optimization_end = NULL, fields = list()) {
+  stages = discard(set_names(list(on_optimization_begin, on_optimizer_before_eval, on_eval_after_design, on_eval_after_benchmark, on_eval_before_archive, on_optimizer_after_eval, on_result, on_optimization_end), c("on_optimization_begin", "on_optimizer_before_eval", "on_eval_after_design", "on_eval_after_benchmark", "on_eval_before_archive", "on_optimizer_after_eval", "on_result",  "on_optimization_end")), is.null)
+  walk(stages, function(stage) assert_function(stage, args = c("callback", "context")))
```

Comment:
So I still think it would be clearer to call the second argument `context_eval` when it is a `ContextEval` and `context_opt` when it is `ContextOptimization`.

Reply:
This does not cost anything and improves clarity I think.

Reply:
I did not do this because the check in line 131 becomes more complicated. But maybe that's not a good argument...